Add option to give every object a nametag or change the nametag text of players

BlockMen opened the pull request:
Default behavior: Only other playernames visible (like currently)

kaadmy commented:
+1, I like this.

Dragonop commented:
+1 This would be a good adition, probably a modder might find it useful.

est31 commented on doc/lua_api.txt:
```diff
@@ -3223,6 +3223,7 @@ Definition tables
         automatic_face_movement_dir = 0.0,
     --  ^ automatically set yaw to movement direction; offset in degrees; false to disable
         backface_culling = true, -- false to disable backface_culling for model
+        nametag = "", -- default empty, for players their player name
```
This should better be "default empty, for players their name is shown if empty"

est31 commented:
What about making `get_nametag_properties` and `set_nametag_properties` valid for all objects, not just players? Now that objects have a nametag too.

est31 commented:
Well, usually I'd say that we should wait for clientside scripting, then the nametag can be adjusted clientside and per-client. Currently the nametag is globally unique.

But I think we shouldn't stop ourselves from this. :+1:  except comments above.

agwilt commented:
+1 Yes, please.

What about just making tags part of the HUD? Or having a config variable for maximum tag rendering distance, which could be set to 0 to remove tags.

BlockMen commented:
@est31 done.
@agwilt For all objects != player it is limited by activity, so if the object is not loaded the nametag is not shown. But i think we should add a limit for nametags per object, yes, but that would be another PR.

est31 commented:
The way this works right now is that there is both `get_nametag_attributes` and the other direct method, but why do we need both?

I'd just ditch one of them, either get_nametag_attributes, and only leave it for players, like before, or the direct method. Yes, it sorta conflicts what I've said earlier, but I didn't mean to add both.

Or manage the nametag text separately from the attributes.

BlockMen commented:
@est31 I did it this way for two reasons:
1. IMO it was wrong to add the nametag color as single feature, not as part of object properties, so i changed this and moved all into object properties.
2. We want keep backwards compatiblity, so set_nametag_attributes must set this specific obj properties which give us the option to remove those functions in the future.

So i suggest to add a warning msg about get/set_name_attributes() being deprecated would be the best solution to move slowly the handling over to properties (only).

est31 commented:
Sounds good.

Best to get @kwolekr into the discussion, he has demanded the attributes thingy.

TeTpaAka commented:
The attributes functions were added, because the original functions I submitted were to special. I didn't thought of adding the nametag to the object properties, which is IMHO a much cleaner solution than to add special functions for every little detail.
I would agree, that the functions get deprecated and are deleted in the long run. (maybe we could replace them with a function that changes the object properties for compatibility).

nerzhul commented on src/content_sao.cpp:
```diff
@@ -857,7 +855,7 @@ std::string PlayerSAO::getClientInitializationData(u16 protocol_version)
 				m_physics_override_jump, m_physics_override_gravity, m_physics_override_sneak,
 				m_physics_override_sneak_glitch)); // 5
-		os << serializeLongString(gob_cmd_update_nametag_attributes(m_nametag_color)); // 6
+		os << serializeString(gob_cmd_update_nametag_attributes(m_prop.nametag_color)); // 6 (GENERIC_CMD_UPDATE_NAMETAG_ATTRIBUTES) : Deprecated, for backwards compatibility only.
```
warning you trigger a protocol break there

sapier commented:
+1 go ahead
